afew/filters/DKIMValidityFilter.py

```diff
@@ -0,0 +1,57 @@
+# coding=utf-8
+from __future__ import print_function, absolute_import, unicode_literals
+
+#
+# Copyright (c) Amadeusz Zolnowski <aidecoe@aidecoe.name>
+#
+# Permission to use, copy, modify, and distribute this software for any
+# purpose with or without fee is hereby granted, provided that the above
+# copyright notice and this permission notice appear in all copies.
+#
+# THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+# WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+# ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+# WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+# ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+# OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+#
+
+import dkim
+
+from .BaseFilter import Filter
+
+
+def verify_dkim(path):
+    '''
+    Verify DKIM signature of an e-mail file.
+
+    :param path: Path to the e-mail file.
+    :returns: Whether DKIM signature is valid or not.
+    '''
+    with open(path, 'rb') as message_file:
+        message_bytes = message_file.read()
+
+    return dkim.verify(message_bytes)
+
+
+class DKIMValidityFilter(Filter):
+    '''
+    Verifies DKIM signature of an e-mail which has DKIM header.
+    '''
+    message = 'Verify DKIM signature'
+    header = 'DKIM-Signature'
+
+    def __init__(self, database, ok_tag='dkim-ok', fail_tag='dkim-fail'):
+        super(DKIMValidityFilter, self).__init__(database)
+        self.ok_tag = ok_tag
+        self.fail_tag = fail_tag
+
+    def handle_message(self, message):
+        if message.get_header(self.header):
+            dkim_ok = all(map(verify_dkim, message.get_filenames()))
```

Comment:
not sure about `message.get_filenames()`. This could yield multiple file names, and we need to open the file in python, which [can easily be done wrongly](https://git.notmuchmail.org/git?p=notmuch;a=blobdiff;f=bindings/python/notmuch/message.py;h=d5b98e4fdfdf5ab19e41fcda61ce0fc57d7b301f;hp=cce377d02a8024608bcf840252faae191d937acb;hb=91fe20cd90ce46bf80c416a5c4451f76b4d69ec5;hpb=073188e6905f1b5a065f696c4d02d0d1b3a7d769).

Is it possible to simply pass `message.get_message_parts()` to `dkim.verify`?

Reply:
This is different. `dkim.verify` takes full raw e-mail message (as bytes, not as string). `get_filenames()` may yield multiple messages that's why for sanity `all()` is used, so we don't have false positive.

Reply:
In other words that's the way to do it and there's no issue here.